### ISSUE TITLE
Writing bgzipped VCF

### ISSUE THREAD

**Issue by MaltheSR** [zaeleus/noodles#10]
I was trying to use noodles (thanks!) to write a VCF file with bgzip compression, which I thought I'd be able to do with a `vcf::Writer<bgzf::Writer<W>>` along the lines of what's happening in the BAM module. 

To illustrate, I tried modifying `noodles-vcf/examples/vcf_write.rs` (changed lines are **_emphasised_**):

<pre>
use std::io;

<b><i>use noodles_bgzf as bgzf;</i></b>
use noodles_vcf::{self as vcf, header::Contig};

fn main() -> Result<(), Box<dyn std::error::Error>> {
    let stdout = io::stdout();
    let handle = stdout.lock();
    <b><i>let bgzf_handle = bgzf::Writer::new(handle);</i></b>
    <b><i>let mut writer = vcf::Writer::new(bgzf_handle);</i></b>

    let header = vcf::Header::builder()
        .add_contig(Contig::new(String::from("sq0")))
        .build();

    writer.write_header(&header)?;

    let record = vcf::Record::builder()
        .set_chromosome("sq0".parse()?)
        .set_position(1)
        .set_reference_bases("A".parse()?)
        .build()?;

    writer.write_record(&record)?;

    Ok(())
}
</pre>

Now, running `cargo run --release --example vcf_write` compiles and runs, but nothing is written to stdout. I tried the same thing with a `std::fs::File` handle instead of a locked stdout, and that likewise compiles and runs, only to produce an empty file. If I explicitly flush the writes, I get a bgzip header and trailer around an uncompressed VCF, and then I wasn't sure how to proceed. Perhaps I'm missing something simpler?

**Comment by zaeleus** [comment]
Hi, @MaltheSR!

Thanks for the report; it's definitely an oversight on my part!

This occurs because `bgzf::Writer` doesn't implement `Drop`, which would (expectedly) flush the current contents of the gzip encoder and append the final BGZF EOF block.

I implemented a fix in a73103af9f8c84675845162a5f599cd2ec6ef81b, so your modified example should work now.

**Comment by MaltheSR** [comment]
Thanks for the quick fix, and for the crate in general - I've been really enjoying reading and using noodles for a project of mine!

While experimenting, I ran into another quirk in the `bgzf` module: when working with a `reader: io::BufReader<bgzf::Reader<R>>`, calls to `reader.get_ref().virtual_position()` break. A little MWE to illustrate:

```
use std::io::{self, BufRead, Read, Write};

use noodles_bgzf as bgzf;

fn log(line: &str, virtual_position: &bgzf::VirtualPosition) {
    println!(
        "Line '{}', compressed offset {}; uncompressed offset {}.",
        line, virtual_position.compressed(), virtual_position.uncompressed()
    );
}

fn main() -> io::Result<()> {
    // Setup

    let input = b"line1\nline2\nline3\nline4\nline5\n";

    let mut writer = bgzf::Writer::new(Vec::new());
    writer.write_all(input)?;
    let data = writer.finish()?;

    // Plain reader (this works as expected)

    println!("Plain reader:");

    let mut reader = bgzf::Reader::new(data.as_slice());
    let mut virtual_position = reader.virtual_position();
    let mut buf = [0u8; 6];

    while let Ok(_) = reader.read_exact(&mut buf) {
        let line = std::str::from_utf8(&buf[0..5]).expect("cannot parse UTF8");
        log(line, &virtual_position);
        virtual_position = reader.virtual_position();
    };

    println!();

    // Buffered reader (this does not work as expected)

    println!("Buffered reader:");

    let mut reader = io::BufReader::new(bgzf::Reader::new(data.as_slice()));

    loop {
        let virtual_position = reader.get_ref().virtual_position();
        let mut line = String::new();
        match reader.read_line(&mut line) {
            Ok(0) => break,
            Err(e) => return Err(e),
            _ => { line.pop(); },
        }
        log(&line, &virtual_position);
    }

    Ok(())
}
```

Running this prints:

```
Plain reader:
Line 'line1', compressed offset 0; uncompressed offset 0.
Line 'line2', compressed offset 0; uncompressed offset 6.
Line 'line3', compressed offset 0; uncompressed offset 12.
Line 'line4', compressed offset 0; uncompressed offset 18.
Line 'line5', compressed offset 0; uncompressed offset 24.

Buffered reader:
Line 'line1', compressed offset 0; uncompressed offset 0.
Line 'line2', compressed offset 46; uncompressed offset 0.
Line 'line3', compressed offset 46; uncompressed offset 0.
Line 'line4', compressed offset 46; uncompressed offset 0.
Line 'line5', compressed offset 46; uncompressed offset 0.
```

I guess this is simply a consequence of the fact that the virtual position will be moved on `BufReader::fill_buf`, since this is where the call to the inner `bgzf::Reader::read` occurs, rather than `BufReader::consume`. Therefore, I thought at first that I was perhaps in unintended territory by going through the `BufReader::get_ref`, but I notice the same thing occurs in `vcf::Reader::virtual_position`, so I thought it might be helpful to mention it here.

The only solution I could think of would be to implement `io::BufRead` on the `bgzf::Reader` directly, to ensure that the position gets updated properly in `consume`. I think that could be made to work, and in some sense it seems kind of natural since the `bgzf::Reader` is already kind of buffered via the calls to `read_block`. I'll be happy to try and help draft a PR if you think this is the way forward, but I thought maybe you might have different ideas for how and whether this should be supported.
